File: C#0.7/Derivative.cs
using System;

namespace MyMath.Calculus
{
    public class Derivative
    {
        private Derivative() { }
        public double easyDerivative(Func<double, double> f, double x, double dx)
        {
            return (f(x + dx) - f(x)) / dx;

        }
        //Sqr NOT because Squared!!! The lose of precision is proportional to d^2x
        public double DerivativeSqr(Func<double, double> f, double x, double dx)
        {
            return (f(x + dx) - f(x-dx)) / (2*dx);

        }
        public double easyLim(Func<double, double> f, double x, double dx)
        {
            double g(double h)
            {
                return (f(x + h) - f(x)) / h;

            }
            return new Limit(g, 0, dx).Apply();

        }
        public double sqrLim(Func<double, double> f, double x, double dx)
        {
            double g(double h)
            {
                return (f(x + h) - f(x-h)) / (2*h);

            }
            return new Limit(g, 0, dx).Apply();

        }

    }

}

File: 2.0.0/JGJAR/libraries/MyMath/calculus/Derivative.java
package JGJAR.libraries.MyMath.calculus;

import JGJAR.libraries.MyMath.DoubleFunction;
import JGJAR.libraries.MyMath.Polynomial;

/**<h1>Derivative class</h1><br>
 * 
 * @author JavaGamesJAR
 * @since MyMath 2.0.0
 * @version 1.0
 */
public class Derivative 
{
	private Derivative()
	{}
	static double eps = 0.000001;
	
	/**Returns derivative of f with easy limit
	 * 
	 * @param f function derivative of you get
	 * @return f'(x)
	 * @since 1.0 (MyMath 2.0.0)
	 */
	public static DoubleFunction easyDeriv(DoubleFunction f)
	{
		return new DoubleFunction() {
			
			@Override
			public double reversedFunction(double x) {
				return 0;
			}
			
			@Override
			public double execute(double x) {
				return (f.execute(x+eps)-f.execute(x))/eps;
			}
		};
		
	}
	
	/**Returns derivative of f with limit, where O &Proportional; h<sup>2</sup>
	 * 
	 * @param f function derivative of you get
	 * @return f'(x)
	 * @since 1.0 (MyMath 2.0.0)
	 */
	public static DoubleFunction sqrDerivative(DoubleFunction f)
	{
		return new DoubleFunction() {
			
			@Override
			public double reversedFunction(double x) {
				return 0;
			}
			
			@Override
			public double execute(double x) {
				return (f.execute(x+eps)-f.execute(x-eps))/(2*eps);
			}
		};
		
	}
	/**Returns derivative using Taylor Series (derives the polynomial of f(x))
	 * 
	 * @param f function derivative of you get
	 * @return f'(x)
	 * @since 1.0 (MyMath 2.0.0)
	 */
	public static DoubleFunction Taylor(DoubleFunction f, int n)
	{
		final Polynomial p = Polynomial.TaylorSeries(f, n);
		p.derivative();
		return new DoubleFunction() {
			
			@Override
			public double reversedFunction(double x) {
				return 0;
			}
			
			@Override
			public double execute(double x) {
				return p.apply(x);
			}
		};
		
	}

}
